Metric View in declarative config need included and excluded Block for attribute_keys

### Describe the bug

I create a metric view using the declarative configuration to include some labels via the `included` directive in the `attributes_key` block.
I haven't needed an `excluded` block so far.

```
meter_provider:
  readers:
    - periodic:
        exporter:
          otlp_http:
            endpoint: http://localhost:4318/v1/metrics
  views:
    - selector:
        instrument_name: http.server.request.duration
      stream:
        attribute_keys:
          included:
            - url.full
            - http.request.method
            - http.response.status_code
            - error.type
            - server.address
            - server.port
            - http.target
            - http.path
            - url.path
```

Now I wanted to test the latest version of the Java agent. With the new version 2.27.0, I now get the following error without an excluded block being set:

```
  [otel.javaagent 2026-04-23 10:18:02:341 +0200] [main] INFO io.opentelemetry.exporter.internal.http.HttpExporter - Calling shutdown() multiple times.
OpenTelemetry Javaagent failed to start
io.opentelemetry.sdk.autoconfigure.spi.ConfigurationException: excluded must not be empty
        at io.opentelemetry.sdk.autoconfigure.IncubatingUtil.toConfigurationException(IncubatingUtil.java:76)
        at io.opentelemetry.sdk.autoconfigure.IncubatingUtil.create(IncubatingUtil.java:69)
        at io.opentelemetry.sdk.autoconfigure.IncubatingUtil.configureFromFile(IncubatingUtil.java:39)
        at io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder.maybeConfigureFromFile(AutoConfiguredOpenTelemetrySdkBuilder.java:598)
        at io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder.buildImpl(AutoConfiguredOpenTelemetrySdkBuilder.java:450)
        at io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder.build(AutoConfiguredOpenTelemetrySdkBuilder.java:431)
        at io.opentelemetry.javaagent.tooling.OpenTelemetryInstaller.installOpenTelemetrySdk(OpenTelemetryInstaller.java:36)
        at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:154)
        at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:101)
        at io.opentelemetry.javaagent.tooling.AgentStarterImpl.start(AgentStarterImpl.java:102)
        at io.opentelemetry.javaagent.bootstrap.AgentInitializer$2.run(AgentInitializer.java:68)
        at io.opentelemetry.javaagent.bootstrap.AgentInitializer$2.run(AgentInitializer.java:61)
        at io.opentelemetry.javaagent.bootstrap.AgentInitializer.execute(AgentInitializer.java:84)
        at io.opentelemetry.javaagent.bootstrap.AgentInitializer.initialize(AgentInitializer.java:60)
        at io.opentelemetry.javaagent.OpenTelemetryAgent.startAgent(OpenTelemetryAgent.java:59)
        at io.opentelemetry.javaagent.OpenTelemetryAgent.premain(OpenTelemetryAgent.java:46)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:560)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:572)
Caused by: io.opentelemetry.javaagent.shaded.io.opentelemetry.api.incubator.config.DeclarativeConfigException: excluded must not be empty
        at io.opentelemetry.sdk.extension.incubator.fileconfig.IncludeExcludeFactory.create(IncludeExcludeFactory.java:32)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.ViewFactory.create(ViewFactory.java:35)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.MeterProviderFactory.lambda$create$1(MeterProviderFactory.java:67)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.MeterProviderFactory.create(MeterProviderFactory.java:61)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.OpenTelemetryConfigurationFactory.create(OpenTelemetryConfigurationFactory.java:87)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.OpenTelemetryConfigurationFactory.create(OpenTelemetryConfigurationFactory.java:25)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.DeclarativeConfiguration.createAndMaybeCleanup(DeclarativeConfiguration.java:279)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.DeclarativeConfiguration.create(DeclarativeConfiguration.java:169)
        at io.opentelemetry.sdk.extension.incubator.fileconfig.DeclarativeConfiguration.create(DeclarativeConfiguration.java:155)
        at io.opentelemetry.sdk.autoconfigure.IncubatingUtil.create(IncubatingUtil.java:66)
        ... 18 more
```


The only workaround is to add any label to an excluded block.
Is this absolutely necessary with the new version? Or should the block actually remain optional?

### Steps to reproduce

- use declarative configuration
- implement a view in meter_provider
- try to implement only a `included` block in attribute_keys
- enable Debug Logging
- start Applikation with javaagent using declarative configuration

### Expected behavior

It should be optional to implement an `excluded` Block, when using an `included` Block

### Actual behavior

Without an `excluded` Block it is not possible to start the agent, if you want to included attributes.

### Javaagent or library instrumentation version

2.27.0

### Environment

**JDK**: openjdk 21.0.10 2026-01-20 LTS
**OS**: Red Hat Enterprise Linux 9.7

### Additional context

_No response_

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

--- a/sdk-extensions/declarative-config/build.gradle.kts
+++ b/sdk-extensions/declarative-config/build.gradle.kts
@@ -107,6 +107,12 @@ jsonSchema2Pojo {
 
   // Append Model as suffix to the generated classes.
   classNameSuffix = "Model"
+
+  // Initialize collection fields to null rather than empty collections so that absent YAML
+  // properties deserialize as null (not present) rather than [] (explicitly empty). This lets
+  // factories distinguish between "user omitted the field" and "user provided an empty list",
+  // which is important for validations like IncludeExcludeFactory.
+  initializeCollections = false
 }
 
 val generateJsonSchema2Pojo = tasks.getByName("generateJsonSchema2Pojo")

--- a/sdk-extensions/declarative-config/src/main/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ComposableRuleBasedSamplerFactory.java
+++ b/sdk-extensions/declarative-config/src/main/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ComposableRuleBasedSamplerFactory.java
@@ -54,12 +54,10 @@ final class ComposableRuleBasedSamplerFactory
           rule -> {
             AttributeMatcher valueMatcher = attributeValuesMatcher(rule.getAttributeValues());
             AttributeMatcher patternMatcher = attributePatternsMatcher(rule.getAttributePatterns());
-            // TODO: should be null when omitted but is empty
             Set<ExperimentalSpanParent> matchingParents =
                 rule.getParent() != null && !rule.getParent().isEmpty()
                     ? new HashSet<>(rule.getParent())
                     : null;
-            // TODO: should be null when omitted but is empty
             Set<SpanKind> matchingSpanKinds =
                 rule.getSpanKinds() != null && !rule.getSpanKinds().isEmpty()
                     ? rule.getSpanKinds().stream()

--- a/sdk-extensions/declarative-config/src/test/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ComposableRuleBasedSamplerFactoryTest.java
+++ b/sdk-extensions/declarative-config/src/test/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ComposableRuleBasedSamplerFactoryTest.java
@@ -135,6 +135,33 @@ class ComposableRuleBasedSamplerFactoryTest {
         Arguments.of(
             new ExperimentalComposableRuleBasedSamplerModel(),
             ComposableSampler.ruleBasedBuilder().build()),
+        // attributePatterns with only included (no excluded) - analogous to
+        // https://github.com/open-telemetry/opentelemetry-java/issues/8337
+        Arguments.of(
+            new ExperimentalComposableRuleBasedSamplerModel()
+                .withRules(
+                    Collections.singletonList(
+                        new ExperimentalComposableRuleBasedSamplerRuleModel()
+                            .withAttributePatterns(
+                                new ExperimentalComposableRuleBasedSamplerRuleAttributePatternsModel()
+                                    .withKey("http.path")
+                                    .withIncluded(Collections.singletonList("/internal/*")))
+                            .withSampler(
+                                new ExperimentalComposableSamplerModel()
+                                    .withAlwaysOn(
+                                        new ExperimentalComposableAlwaysOnSamplerModel())))),
+            ComposableSampler.ruleBasedBuilder()
+                .add(
+                    new DeclarativeConfigSamplingPredicate(
+                        null,
+                        new AttributeMatcher(
+                            "http.path",
+                            IncludeExcludePredicate.createPatternMatching(
+                                Collections.singletonList("/internal/*"), null)),
+                        null,
+                        null),
+                    ComposableSampler.alwaysOn())
+                .build()),
         // Recreate example
         Arguments.of(
             new ExperimentalComposableRuleBasedSamplerModel()

--- a/sdk-extensions/declarative-config/src/test/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ViewFactoryTest.java
+++ b/sdk-extensions/declarative-config/src/test/java/io/opentelemetry/sdk/autoconfigure/declarativeconfig/ViewFactoryTest.java
@@ -18,56 +18,65 @@ import io.opentelemetry.sdk.metrics.ExplicitBucketHistogramOptions;
 import io.opentelemetry.sdk.metrics.View;
 import java.util.Arrays;
 import java.util.Collections;
-import org.junit.jupiter.api.Test;
+import java.util.stream.Stream;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
 class ViewFactoryTest {
 
-  @Test
-  void create_Defaults() {
-    View expectedView = View.builder().build();
-
-    View view =
-        ViewFactory.getInstance()
-            .create(
-                new ViewStreamModel().withAttributeKeys(null),
-                mock(DeclarativeConfigContext.class));
-
+  @ParameterizedTest
+  @MethodSource("createArguments")
+  void create(ViewStreamModel model, View expectedView) {
+    View view = ViewFactory.getInstance().create(model, mock(DeclarativeConfigContext.class));
     assertThat(view.toString()).isEqualTo(expectedView.toString());
   }
 
-  @Test
-  void create() {
-    View expectedView =
-        View.builder()
-            .setName("name")
-            .setDescription("description")
-            .setAttributeFilter(
-                IncludeExcludePredicate.createPatternMatching(
-                    Arrays.asList("foo", "bar"), Collections.singletonList("baz")))
-            .setAggregation(
-                Aggregation.explicitBucketHistogram(
-                    ExplicitBucketHistogramOptions.builder()
-                        .setBucketBoundaries(Arrays.asList(1.0, 2.0))
-                        .build()))
-            .build();
-
-    View view =
-        ViewFactory.getInstance()
-            .create(
-                new ViewStreamModel()
-                    .withName("name")
-                    .withDescription("description")
-                    .withAttributeKeys(
-                        new IncludeExcludeModel()
-                            .withIncluded(Arrays.asList("foo", "bar"))
-                            .withExcluded(Collections.singletonList("baz")))
-                    .withAggregation(
-                        new AggregationModel()
-                            .withExplicitBucketHistogram(
-                                new ExplicitBucketHistogramAggregationModel()
-                                    .withBoundaries(Arrays.asList(1.0, 2.0)))),
-                mock(DeclarativeConfigContext.class));
-
-    assertThat(view.toString()).isEqualTo(expectedView.toString());
+  private static Stream<Arguments> createArguments() {
+    return Stream.of(
+        // defaults
+        Arguments.of(new ViewStreamModel().withAttributeKeys(null), View.builder().build()),
+        // attribute_keys with only included (no excluded) - reproduces
+        // https://github.com/open-telemetry/opentelemetry-java/issues/8337
+        Arguments.of(
+            new ViewStreamModel()
+                .withAttributeKeys(
+                    new IncludeExcludeModel()
+                        .withIncluded(
+                            Arrays.asList(
+                                "url.full", "http.request.method", "http.response.status_code"))),
+            View.builder()
+                .setAttributeFilter(
+                    IncludeExcludePredicate.createPatternMatching(
+                        Arrays.asList(
+                            "url.full", "http.request.method", "http.response.status_code"),
+                        null))
+                .build()),
+        // full configuration
+        Arguments.of(
+            new ViewStreamModel()
+                .withName("name")
+                .withDescription("description")
+                .withAttributeKeys(
+                    new IncludeExcludeModel()
+                        .withIncluded(Arrays.asList("foo", "bar"))
+                        .withExcluded(Collections.singletonList("baz")))
+                .withAggregation(
+                    new AggregationModel()
+                        .withExplicitBucketHistogram(
+                            new ExplicitBucketHistogramAggregationModel()
+                                .withBoundaries(Arrays.asList(1.0, 2.0)))),
+            View.builder()
+                .setName("name")
+                .setDescription("description")
+                .setAttributeFilter(
+                    IncludeExcludePredicate.createPatternMatching(
+                        Arrays.asList("foo", "bar"), Collections.singletonList("baz")))
+                .setAggregation(
+                    Aggregation.explicitBucketHistogram(
+                        ExplicitBucketHistogramOptions.builder()
+                            .setBucketBoundaries(Arrays.asList(1.0, 2.0))
+                            .build()))
+                .build()));
   }
 }

